drivers/uaol/uaol_intel_adsp.c

```diff
@@ -0,0 +1,861 @@
+		uint8_t recipient : 5;
+		uint8_t type      : 2;
+		uint8_t direction : 1;
+		uint8_t bRequest;
+		uint16_t wValue;
+		uint16_t wIndex;
+		uint16_t wLength;
+	} __packed __aligned(1) payload;
+} __packed __aligned(4);
+
+/* xHCI response message */
+struct xhci_resp_msg {
+	uint8_t type;
+	uint16_t length;
+	uint8_t completion_code;
+} __packed __aligned(4);
+
+/* xHCI EP_TABLE_ENTRY message */
+struct xhci_ep_table_entry_msg {
+	uint8_t type;
+	uint16_t length;
+	struct {
+		uint16_t usb_ep_address     : 5;
+		uint16_t device_slot_number : 8;
+		uint16_t split_ep           : 1;
+		uint16_t rsvd0              : 1;
+		uint16_t valid              : 1;
+		uint8_t sio_pin_number;
+		uint8_t rsvd;
+	} __packed __aligned(1) payload;
+} __packed __aligned(4);
+
+/* xHCI HH_TIMESTAMP message */
+struct xhci_hh_timestamp_msg {
+	uint8_t type;
+	uint16_t length;
+	struct {
+		uint16_t cmfb  : 13;
+		uint16_t rsvd0 : 3;
+		uint16_t cmfi  : 14;
+		uint16_t rsvd1 : 2;
+		uint64_t global_time;
+	} __packed __aligned(1) payload;
+} __packed __aligned(4);
+
+/*
+ * Set power state (up/down) for UAOL link.
+ */
+static int uaol_intel_adsp_set_power(const struct device *dev, bool power)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+	union UAOLCTL ctl;
+	uint64_t start_time;
+	uint64_t timeout = k_us_to_cyc_ceil64(UAOL_POWER_CHANGE_TIMEOUT_USEC);
+	uint32_t link_mask = BIT(dp->link);
+
+	ctl.full = sys_read32(UAOLCTL_ADDR(dp));
+	if ((ctl.part.cpa & link_mask) != (ctl.part.spa & link_mask)) {
+		return -EBUSY;
+	}
+	ctl.part.spa = (power) ? (ctl.part.spa | link_mask) : (ctl.part.spa & ~link_mask);
+	sys_write32(ctl.full, UAOLCTL_ADDR(dp));
+
+	start_time = sys_clock_cycle_get_64();
+	while (1) {
+		ctl.full = sys_read32(UAOLCTL_ADDR(dp));
+		if ((ctl.part.cpa & link_mask) == (ctl.part.spa & link_mask)) {
+			break;
+		}
+		if (sys_clock_cycle_get_64() - start_time > timeout) {
+			return -ETIMEDOUT;
+		}
+		k_busy_wait(1);
+	}
+
+	return 0;
+}
+
+/*
+ * Read HW capabilities for UAOL link.
+ */
+static void uaol_intel_adsp_read_capabilities(const struct device *dev,
+					      struct uaol_capabilities *caps)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+	union UAOLxPCMSCAP pcmscap;
+	union UAOLxOPC opc;
+	union UAOLxIPC ipc;
+
+	pcmscap.full = sys_read16(UAOLxPCMSCAP_ADDR(dp));
+	opc.full = sys_read16(UAOLxOPC_ADDR(dp));
+	ipc.full = sys_read16(UAOLxIPC_ADDR(dp));
+
+	caps->input_streams = pcmscap.part.iss;
+	caps->output_streams = pcmscap.part.oss;
+	caps->bidirectional_streams = pcmscap.part.bss;
+	caps->max_tx_fifo_size = opc.part.opc;
+	caps->max_rx_fifo_size = ipc.part.ipc;
+}
+
+/*
+ * Program the target destination ID of the xHCI Controller for UAOL link.
+ */
+static void uaol_intel_adsp_program_bdf(const struct device *dev, uint8_t bus,
+					uint8_t device, uint8_t function)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+	union UAOLxTBDF tbdf = {0};
+
+	tbdf.part.busn = bus;
+	tbdf.part.devn = device;
+	tbdf.part.fncn = function;
```

Comment:
Style: struct initializers are your friend:

```c
union UAOLxTBDF tbdf = {
      .part.busn = bus,
      .part.devn = device,
      .part.fncn = function,
};
```

Reply:
Also same comment from elsewhere that single-register inlines are IMHO a waste of space.

Reply:
Fixed this occurrence.